packages/back-end/src/models/VisualChangesetModel.ts

```diff
@@ -0,0 +1,93 @@
+import omit from "lodash/omit";
+import mongoose from "mongoose";
+import { ApiVisualChangeset } from "../../types/openapi";
+import { VisualChangesetInterface } from "../../types/visual-changeset";
+
+/**
+ * VisualChangeset is a collection of visual changes that are grouped together
+ * by a single url target. They are many-to-one with Experiments.
+ */
+const visualChangesetSchema = new mongoose.Schema({
+  id: {
+    type: String,
+    unique: true,
+  },
+  organization: {
+    type: String,
+    index: true,
+  },
+  urlPattern: String,
+  editorUrl: String,
+  experiment: {
+    type: String,
+    index: true,
+  },
+  // VisualChanges are associated with one of the variations of the experiment
+  // associated with the VisualChangeset
+  visualChanges: [
+    {
+      id: String,
+      description: String,
+      css: String,
+      variation: {
+        type: String,
+        index: true,
+      },
+      domMutations: [
+        {
+          selector: String,
+          action: ["append", "set", "remove"],
+          attribute: String,
+          value: String,
+        },
+      ],
+    },
+  ],
```

Comment:
If I'm not mistaken, and feel free to correct me if I'm wrong, I don't think schema validation works for nested definitions, which is why I implemented a validator in the models I defined. You can see an example [here](https://github.com/growthbook/growthbook/blob/f92260cac9fa2fbbc857561963d62f33bcc86188/packages/back-end/src/models/EventWebhookModel.ts#L44-L55).

An easy way to test validation is by adding this new model to the [REPL file](https://github.com/growthbook/growthbook/blob/f92260cac9fa2fbbc857561963d62f33bcc86188/packages/back-end/repl.ts) and running `yarn repl` in the backend package and trying to evaluate invalid schema definitions in the console.

Reply:
Nice, this is a good reference, thank you.

I made some changes based on the reference you provided. The nested collections defined (`visualChanges` and `domMutations`) both can be empty in a valid state, so a minimum length isn't necessary.

Nested validation does seem to work out of the box with regard to validating against `required` setting and checking against enum lists. Here is an example where I tried to create a `VisualChangeset` with a nested `domMutation` set with an invalid enum value

Input

```
 var y = await x.create({ id: '2343', organization: '123', urlPattern: '123', editorUrl: '123', experiment: '123', visualChanges: [{ id: '222', variation: 'abc', domMutations: [{ selector: 'true', action: 'not-valid', attribute: 'class' }] }] });
```

Output
<img width="861" alt="Screenshot 2023-03-03 at 11 24 08 AM" src="https://user-images.githubusercontent.com/2374625/222773757-7d06a6f1-d6b9-44fa-b279-4ae83d405b7e.png">

The above validation should suffice IMO, but let me know if you disagree.

Reply:
Ok great, glad it's working, thanks for verifying. My understanding is that required checking didn't work for nested properties in Mongoose unless you define the nested object as a schema ([reference](https://mongoosejs.com/docs/validation.html#required-validators-on-nested-objects)). If you tested it and it works, then that's fine.